test/e2e/tests/test_role.py

```diff
@@ -212,8 +213,93 @@ def test_crud(self, simple_role):
 
         cr = k8s.get_resource(ref)
         assert cr is not None
-        assert 'spec' in cr
-        assert 'inlinePolicies' not in cr['spec']
+        assert "spec" in cr
+        assert "inlinePolicies" not in cr["spec"]
 
         latest_inline_policies = role.get_inline_policies(role_name)
         assert len(latest_inline_policies) == 0
+
+
+        # assumeRolePolicyDocument tests:
+        # Assert that the policy is as expected.
+        # Modify it.
+        # Delete it
+
+        assume_role_policy_doc = '''
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Principal": {
+                "Service": ["ec2.amazonaws.com"]
+            },
+            "Action": ["sts:AssumeRole"]
+        }
+    ]
+}
+'''
+
+        cr = k8s.get_resource(ref)
+        assert cr is not None
+        assert 'spec' in cr
+        assert 'assumeRolePolicyDocument' in cr['spec']
+
+        assume_role_policy_as_obj = json.loads(assume_role_policy_doc)
+        k8s_assume_role_policy = json.loads(cr['spec']['assumeRolePolicyDocument'])
+        assert assume_role_policy_as_obj == k8s_assume_role_policy
+
+        assume_role_policy_to_deny_doc = '''
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Deny",
+            "Principal": {
+                "Service": ["ec2.amazonaws.com"]
+            },
+            "Action": ["sts:AssumeRole"]
+        }
+    ]
+}
+'''
+
+        updates = {
+            'spec': {
+                'assumeRolePolicyDocument': assume_role_policy_to_deny_doc,
+            }
+        }
+
+        k8s.patch_custom_resource(ref, updates)
+        time.sleep(MODIFY_WAIT_AFTER_SECONDS)
+
+        cr = k8s.get_resource(ref)
+        assert cr is not None
+        assert 'spec' in cr
+        assert 'assumeRolePolicyDocument' in cr['spec']
+
+        assume_role_policy_deny_obj = json.loads(assume_role_policy_to_deny_doc)
+        k8s_assume_role_policy_deny = json.loads(cr['spec']['assumeRolePolicyDocument'])
+        assert assume_role_policy_deny_obj == k8s_assume_role_policy_deny
+
+        # AWS slightly modifies the JSON structure underneath us here, so the documents
+        # are not identical. Instead, we can ensure that the change we made is reflected.
+        latest_assume_role_policy_doc = role.get_assume_role_policy(role_name)
+        assert latest_assume_role_policy_doc['Statement'][0]['Effect'] == k8s_assume_role_policy_deny['Statement'][0]['Effect']
```

Comment:
@jaypipes The last hurdle on this one, I can't seem to figure out why the document isn't being updated on the AWS side to then be read as an updated K8s object. No doubt I'm missing something very obvious, I'll pick it back up later on this evening 😊 

Output from `make kind-test SERVICE=iam` 👇 
```
>           assert latest_assume_role_policy_doc['Statement'][0]['Effect'] == k8s_assume_role_policy_deny['Statement'][0]['Effect']
E           AssertionError: assert 'Allow' == 'Deny'
E             - Deny
E             + Allow
```

Have I missed something on registering the new hook here? So that it is not picking up the change and running through the `putAssumeRolePolicy` func to change the document.

Reply:
Looks like the API rejected the new policy:

```
2023-04-04T09:50:44.212Z	INFO	ackrt	desired resource state has changed	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9, "diff": [{"Path":{"Parts":["Spec","AssumeRolePolicyDocument"]},"A":"\n{\n    \"Version\": \"2012-10-17\",\n    \"Statement\": [\n        {\n            \"Effect\": \"Deny\",\n            \"Principal\": {\n                \"Service\": [\"ec2.amazonaws.com\"]\n            },\n            \"Action\": [\"sts:AssumeRole\"]\n        }\n    ]\n}\n","B":"{\"Version\":\"2012-10-17\",\"Statement\":[{\"Effect\":\"Allow\",\"Principal\":{\"Service\":\"ec2.amazonaws.com\"},\"Action\":\"sts:AssumeRole\"}]}"},{"Path":{"Parts":["Spec","Description"]},"A":null,"B":"a simple role"},{"Path":{"Parts":["Spec","InlinePolicies"]},"A":null,"B":{}}]}
2023-04-04T09:50:44.212Z	DEBUG	ackrt	>>> rm.Update	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9}
2023-04-04T09:50:44.212Z	DEBUG	ackrt	>>>> rm.sdkUpdate	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9}
2023-04-04T09:50:44.212Z	DEBUG	ackrt	>>>>> rm.syncInlinePolicies	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9}
2023-04-04T09:50:44.212Z	DEBUG	ackrt	<<<<< rm.syncInlinePolicies	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9}
2023-04-04T09:50:44.212Z	DEBUG	ackrt	>>>>> rm.putAssumeRolePolicy	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9}
2023-04-04T09:50:44.355Z	DEBUG	ackrt	<<<<< rm.putAssumeRolePolicy	{"account": "881878278316", "role": "", "region": "us-west-2", "kind": "Role", "namespace": "default", "name": "my-simple-role-0yfzzmequ", "is_adopted": false, "generation": 9, "error": "MalformedPolicyDocument: JSON strings must not have leading spaces\n\tstatus code: 400, request id: 4f941417-7e5e-4f5b-a6fc-a10f147e8398
```
logs from https://prow.ack.aws.dev/view/s3/ack-prow-logs/pr-logs/pull/aws-controllers-k8s_iam-controller/74/iam-kind-e2e/1643186964129648640

@jdockerty To run e2e tests on this PR you can either run `/test all` or make your PR ready for review :)